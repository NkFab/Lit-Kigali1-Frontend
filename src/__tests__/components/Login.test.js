import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import {
  LoginComponent as Login,
  mapDispatchToProps,
  mapStateToProps,
} from '../../components/Auth/Login';

let wrapper;
let store;
const mockFn = jest.fn();

const props = {
  login: jest.fn().mockImplementation(() => Promise.resolve({ status: 200 })),
  handleInput: mockFn,
  validate: jest.fn().mockImplementation(() => Promise.resolve({ message: 'Ok' })),
  history: { push: mockFn },
};

const mockStore = configureMockStore();

describe('<Login />', () => {
  test('should render the <Login />', () => {
    const renderedValue = renderer
      .create(
        <Router>
          <Login {...props} />
        </Router>,
      )
      .toJSON();
    expect(renderedValue).toMatchSnapshot();
  });

  test('Should call handleInput', () => {
    const event = { target: { value: 'email@email.com' } };
    wrapper = shallow(<Login {...props} />);
    wrapper.find('#username').simulate('change', event);
    expect(props.handleInput).toHaveBeenCalled();
  });

  test('should add `loading` class to submit button', () => {
    wrapper.setProps({ submitting: true });
    expect(wrapper.find('Button').hasClass('loading'));
  });

  describe('Clicking submit button', () => {
    beforeEach(() => {
      store = mockStore({});
      wrapper = mount(
        <Router>
          <Login store={store} {...props} />
        </Router>,
      );
    });

    test('should call submit function', () => {
      wrapper.setProps({
        credentials: {
          username: 'christian',
          password: '123456',
        },
      });
      wrapper.find('.button').simulate('click');
      expect(props.validate).toHaveBeenCalled();
    });
  });

  describe('actions creators', () => {
    test('should call handleInput action', () => {
      const dispatch = jest.fn();
      const payload = { field: 'username', value: 'email@email.com' };
      const expectedActions = {
        type: 'HANDLE_LOGIN_INPUT',
        payload,
      };
      mapDispatchToProps(dispatch).handleInput(payload);
      expect(dispatch.mock.calls[0][0]).toEqual(expectedActions);
    });

    test('should call login action', () => {
      const dispatch = jest.fn();
      const payload = { username: 'email@email.com', password: '123456' };
      mapDispatchToProps(dispatch).login(payload);
      expect(dispatch.mock.calls[0][0]).toBeDefined();
    });

    test('should call validate action', () => {
      const dispatch = jest.fn();
      const payload = { username: 'email@email.com', password: '123456' };
      mapDispatchToProps(dispatch).validate(payload);
      expect(dispatch.mock.calls[0][0]).toBeDefined();
    });
  });

  describe('reducers', () => {
    test('should call handleInput action', () => {
      const initialState = {
        login: {
          submitting: false,
          credentials: {
            username: '',
            password: '',
          },
        },
        currentUser: { isLoggedIn: false },
      };
      const expectedState = {
        credentials: { username: '', password: '' },
        error: undefined,
        submitting: false,
        isLoggedIn: false,
      };
      const state = mapStateToProps(initialState);
      expect(state).toEqual(expectedState);
    });
  });
});
