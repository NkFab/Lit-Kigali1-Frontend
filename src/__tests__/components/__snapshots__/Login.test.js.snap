// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render the <Login /> 1`] = `
<div>
  <form
    className="bg-primary-light login-form"
    id="login"
  >
    <div
      className="color-primary to-center"
    >
      <h2>
        LOG IN
      </h2>
    </div>
    <div
      className="form-errors"
      data-test="form-errors"
    >
      
    </div>
    <div
      className="input primary"
    >
      <i
        className="fa fa-user"
      />
      <input
        className="bg-primary-light"
        name="username"
        onChange={[Function]}
        placeholder="Username / Email"
        type="text"
        value=""
      />
      <div
        className="errors"
      />
    </div>
    <div
      className="input primary"
    >
      <i
        className="fa fa-lock"
      />
      <input
        className="bg-primary-light"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <div
        className="errors"
      />
    </div>
    <div
      className="align-right"
      id="forget-psswd"
    >
      <a
        href="/forgot-password"
        onClick={[Function]}
      >
        Forgot Password?
      </a>
    </div>
    <button
      className="button primary color-white "
      disabled=""
      onClick={[Function]}
      type="button"
    >
      Log In
    </button>
    <div
      className="icon-group"
    >
      <div
        data-test="fb-container"
        id="fb"
      >
        <a
          data-test="fb-a"
          href="/#"
        >
          <img
            alt="fb-logo"
            data-test="fb-img"
            src="test-file-stub"
          />
        </a>
      </div>
      <div
        data-test="twbs-container"
        id="twbs"
      >
        <a
          data-test="twbs-a"
          href="/#"
        >
          <img
            alt="twbs-logo"
            data-test="twbs-img"
            src="test-file-stub"
          />
        </a>
      </div>
      <div
        data-test="gl-container"
        id="gl"
      >
        <a
          data-test="gl-a"
          href="/#"
        >
          <img
            alt="gl-logo"
            data-test="gl-img"
            src="test-file-stub"
          />
        </a>
      </div>
    </div>
    <div
      className="to-center"
      id="form-link"
    >
      <span>
        Not a member?
      </span>
      <a
        href="#signup"
      >
        Sign Up
      </a>
    </div>
  </form>
</div>
`;
